WorkloadTools/Consumer/Replay/ReplayWorker.cs

```diff
@@ -270,33 +273,34 @@ public void ExecuteCommand(ReplayCommand command, int failRetryCount = 0, int ti
                     conn.ChangeDatabase(command.Database);
                 }
 
-                // if the command comes with a replay offset, do it now
-                // the offset in milliseconds is set in
-                // FileWorkloadListener
+                // If the command comes with a replay offset, evaluate it now.
+                // The offset in milliseconds is set in FileWorkloadListener.
                 // The other listeners do not set this value, as they
                 // already come with the original timing
                 if (command.ReplayOffset > 0)
                 {
-                    // I am using 7 here as an average compensation for sleep
-                    // fluctuations due to Windows preemptive scheduling
-                    while((DateTime.Now - StartTime).TotalMilliseconds < command.ReplayOffset - 7)
+                    double delay = command.ReplayOffset - (DateTime.Now - StartTime).TotalMilliseconds;
+                    // Delay execution only if necessary
+                    if (delay > 0)
                     {
-                        // Thread.Sleep will not sleep exactly 1 millisecond.
-                        // It will yield the current thread and put it back 
-                        // in the runnable queue once the sleep delay has expired.
-                        // This means that the actual sleep time before the 
-                        // current thread gains back control can be much higher 
-                        // (15 milliseconds or more)
-                        // However we do not need to be super precise here: 
-                        // each command has a requested offset from the beginning
+                        // Each command has a requested offset from the beginning
                         // of the workload and this class does its best to respect it.
                         // If the previous commands take longer in the target environment
                         // the offset cannot be respected and the command will execute
                         // without further waits, but there is no way to recover 
                         // the delay that has built up to that point.
-                        Thread.Sleep(1);
+
+                        var stopwatch = Stopwatch.StartNew();
+                        while (stopwatch.Elapsed.TotalMilliseconds < delay)
+                        {
+                            // If we're getting close to the command time then don't sleep,
+                            // we need higher accuracy to keep workers on time so stay in a tigher loop
+                            if ((delay - stopwatch.Elapsed.TotalMilliseconds) < ReplayOffsetSleepThresholdMs)
+                                continue; // Immediately re-evaluate
```

Comment:
I'm a bit concerned about CPU usage in this case. Wouldn't it be safer to use `_spinWait.SpinOnce()` here? It would still be a busy wait and would not give up the scheduler. Thoughts?

Reply:
Thank you very much for contributing to WorkloadTools! I greatly appreciate your contributions and the time you took to go through the code and make it better.

Reply:
> I'm a bit concerned about CPU usage in this case. Wouldn't it be safer to use `_spinWait.SpinOnce()` here? It would still be a busy wait and would not give up the scheduler. Thoughts?

Yes indeed this will lead to higher cpu usage, it was worth the trade off because all commands were offset exactly when they should be. In my own testing it was tolerable (~10%) but your own testing may be different with a busier workload.  A spin would likely be more accurate than sleeping, so you could spin if within the _ReplayOffsetSleepThresholdMs_, and sleep for a longer period of time (ie. _ReplayOffsetSleepThresholdMs/2_) when outside of it.